tendermint/src/lite/verifier.rs

```diff
@@ -284,42 +239,100 @@ pub fn verify_header<TS, SH, C, L, S, R>(
+    // Note we only check for expiry once in this
+    // verify_and_update_bisection function since we assume the
+    // time is passed in and we don't access a clock internall.
+    // Thus the trust_period must be long enough to incorporate the
+    // expected time to complete this function.
+    let trusted_state = store.get(Height::from(0))?;
+    let trusted_sh = trusted_state.last_header();
+    is_within_trust_period(trusted_sh.header(), trusting_period, now)?;
+
+    // TODO: consider fetching the header we're trying to sync to and
+    // checking that it's time is less then `now + X` for some small X.
+    // If not, it means that either our local clock is really slow
+    // or the blockchains BFT time is really wrong.
+    // In either case, we should probably raise an error.
+    // Note this would be stronger than checking that the untrusted
+    // header is within the trusting period, as it could still diverge
+    // significantly from `now`.
+
+    // inner recursive function which assumes
+    // trusting_period check is already done.
+    _verify_and_update_bisection(untrusted_height, trust_threshold, req, store)
+}
+
+// inner recursive function for verify_and_update_bisection.
+// see that function's docs.
+fn _verify_and_update_bisection<TS, SH, C, L, R, S>(
+    untrusted_height: Height,
+    trust_threshold: &L,
+    req: &R,
+    store: &mut S,
+) -> Result<(), Error>
+where
+    TS: TrustedState<LastHeader = SH, ValidatorSet = C::ValidatorSet>,
     SH: SignedHeader<Commit = C>,
+    C: Commit,
+    L: TrustThreshold,
+    R: Requester<SignedHeader = SH, ValidatorSet = C::ValidatorSet>,
+    S: Store<TrustedState = TS>,
 {
-    // Check if we already trusted a header at the given height and it didn't expire:
-    if let Ok(ts2) = store.get(height) {
-        is_within_trust_period(ts2.last_header().header(), trusting_period, now)?
-    }
-
-    // We haven't trusted a header at given height yet. Request it:
-    let sh2 = req.signed_header(height)?;
-    let sh2_vals = req.validator_set(height)?;
-    verify(&sh2, &sh2_vals)?;
-    is_within_trust_period(sh2.header(), trusting_period, now)?;
-
-    // Get the highest trusted header with height lower than sh2's.
-    let sh1_trusted = store.get_smaller_or_equal(height)?;
-    can_trust_bisection(
-        &sh1_trusted,
-        &sh2,
+    // this get is redundant the first time.
+    // TODO: possibly refactor so this func takes and returns
+    // trusted_state.
+    let trusted_state = store.get(Height::from(0))?;
+    let trusted_sh = trusted_state.last_header();
+
+    // fetch the header and vals for the new height
+    let untrusted_sh = &req.signed_header(untrusted_height)?;
+    let untrusted_vals = &req.validator_set(untrusted_height)?;
+    let untrusted_next_vals = &req.validator_set(untrusted_height.increment())?;
+
+    // check if we can skip to this height and if it verifies.
+    match verify_single(
```

Comment:
...second issue: looks like we call verify_single() multiple times for the untrusted signed header. We expect this to be called with different (higher) trusted singed header but still we do the untrusted header verification (independent of the trusted one) multiple times and this is expensive.

Reply:
We're only doing the validation, which is just hash checks. the hashes can be memoized so this would be cheap equality testing.

We do end up running verify_commit_trusting multiple times which is expensive and it would be interesting to see how we can optimize that but it will start to mess with our traits again